src/PolyExecution/PolyCommand.php

```diff
@@ -0,0 +1,182 @@
+     * See https://symfony.com/doc/current/console/coloring.html
+     *
+     * @return int 0 if everything went fine, or an exit code
+     */
+    abstract public function run(InputInterface $input, PolyOutput $output): int;
+
+    /**
+     * Get the name of this command.
+     */
+    public static function getName(): string
+    {
+        return static::$commandName;
+    }
+
+    /**
+     * Get the description of this command. Includes unparsed symfony/console styling.
+     */
+    public static function getDescription(): string
+    {
+        return _t(static::class . '.description', static::$description);
+    }
+
+    /**
+     * Return additional help context to avoid an overly long description.
+     */
+    public static function getHelp(): string
+    {
+        return '';
+    }
+
+    /**
+     * Get input options that can be passed into the command.
+     *
+     * In CLI execution these will be passed as flags.
+     * In HTTP execution these will be passed in the query string.
+     *
+     * @return array<InputOption>
+     */
+    public function getOptions(): array
+    {
+        return [];
+    }
+
+    public function getOptionsForTemplate(): array
+    {
+        $formatter = HtmlOutputFormatter::create();
+        $forTemplate = [];
+        foreach ($this->getOptions() as $option) {
+            $default = $option->getDefault();
+            if (is_bool($default)) {
+                // Use 1/0 for boolean, since that's what you'd pass in the query string
+                $default = $default ? '1' : '0';
+            }
+            if (is_array($default)) {
+                $default = implode(',', $default);
+            }
+            $forTemplate[] = [
+                'Name' => $option->getName(),
+                'Description' => DBField::create_field('HTMLText', $formatter->format($option->getDescription())),
+                'Default' => $default,
+            ];
+        }
+        return $forTemplate;
+    }
+
+    /**
+     * Check whether this command can be run in CLI via sake
+     */
+    public static function canRunInCli(): bool
+    {
+        static::checkPrerequisites();
+        return Director::isDev()
+            || static::config()->get('can_run_in_cli')
+            || DevelopmentAdmin::config()->get('allow_all_cli');
+    }
+
+    /**
+     * Check whether this command can be run in the browser via a web request
+     */
+    public static function canRunInBrowser(): bool
+    {
+        static::checkPrerequisites();
+        // Can always run in browser in dev mode
+        if (Director::isDev()) {
+            return true;
+        }
+        if (!static::config()->get('can_run_in_browser')) {
+            return false;
+        }
+        // Check permissions if there are any
+        $permissions = static::config()->get('permissions_for_browser_execution');
+        if ($permissions) {
+            return Permission::check($permissions);
```

Comment:
By default this will do an 'any' check, which is different from how LeftAndMain::required_permission_codes does it which does it as an 'all' (not that any of our LeftAndMain subclasses have multiple codes)

Is that what we want?

Reply:
This is copied over from the now deprecated `init_permissions` config property that existed on the controllers.
The idea is that you can add finer grain control (admins have access to all commands, people who have "all dev" have access to al dev commands, and people with specific permissions for individual commands have access to just those specific commands).